### ISSUE TITLE
Critical dependency error when using API in Next.js

### ISSUE THREAD

**Issue by LudoLogical** [prettier/prettier#15706]
**Environments:**

- Prettier Version: 3.1.0
- Running Prettier via: Node.js API, Browser API
- Runtime: Next.js 14.0.3 atop Node.js 20.9.0
- Operating System: Linux (WSL)
- Prettier plugins (if any): `prettier/plugins/babel` and `prettier/plugins/estree`

**Steps to reproduce:**

1. `npx create-next-app@latest` (Yes to TypeScript, ESLint, TailwindCSS, `src/`, and App Router, no to custom import alias)
2. `cd` into the app
3. `npm install prettier`
4. add `"use client";` to the top of `src/app/page.tsx`
5. insert the code snippet below inside `function` `Home`, just before the `return` statement
6. `npm run dev`
7. navigate to `http://localhost:3000` in a web browser

Note that the same problem can be encountered by creating a `route.ts` file somewhere like `src/app/test-route`, attempting to use the `prettier.format()` call from the code snippet below in a [Route Handler](https://nextjs.org/docs/app/building-your-application/routing/route-handlers) endpoint there, and then sending an HTTP request to that endpoint. Thus, the issue can be encountered both on the client and on the server.

Code snippet:
```tsx
useEffect(() => {
    const makePrettier = async () => {
      console.log(
        await prettier.format(
          "export default function Test(){return(<p>hi</p>);}",
          { parser: "babel", plugins: [babelPlugin, estreePlugin] }
        )
      );
    };
    makePrettier();
  }, []);
```

**Expected behavior:**

The following output appears in the terminal:
```
   ▲ Next.js 14.0.3
   - Local:        http://localhost:3000

 ✓ Ready in 2.6s
 ○ Compiling / ...
 ✓ Compiled / in 1756ms (492 modules)
 ✓ Compiled in 593ms (241 modules)
```

**Actual behavior:**

The following output appears in the terminal:
```
   ▲ Next.js 14.0.3
   - Local:        http://localhost:3000

 ✓ Ready in 2.8s
 ○ Compiling / ...
 ⚠ ./node_modules/prettier/index.mjs
Critical dependency: the request of a dependency is an expression

Import trace for requested module:
./node_modules/prettier/index.mjs
./src/app/page.tsx
 ⚠ ./node_modules/prettier/index.mjs
Critical dependency: the request of a dependency is an expression

Import trace for requested module:
./node_modules/prettier/index.mjs
./src/app/page.tsx
```

**Comment by kachkaev** [comment]
👋 @LudoLogical! I believe that this problem is related to how Next.js bundles code, so I’d suggest discussing it in the Next.js community. There are [quite a few issues](https://github.com/search?q=repo%3Avercel%2Fnext.js+Critical+dependency%3A+the+request+of+a+dependency+is+an+expression&type=issues) mentioning `Critical dependency: the request of a dependency is an expression` there.

For server components and routes you can try adding [`serverComponentsExternalPackages: ['prettier']`](https://nextjs.org/docs/app/api-reference/next-config-js/serverComponentsExternalPackages) to your `next.config.js` file. Not sure what to do with usage in `useEffect` though. [Prettier Playground](https://prettier.io/playground/) runs Prettier on the client without warnings, so if the same does not work in Next.js, I guess the cause is external.

If you discover a specific issue in Prettier `package.json` or code, feel free to share your finding and we’ll reopen this issue.

**Comment by LudoLogical** [comment]
👋 @kachkaev! Thank you very much for the quick response, resources, and suggestions. I've done some more digging and discovered the following:

1. A nice feature of Next.js is that, when it bundles a web app, it automatically bundles any dependencies of that web app as well. Like most meta-frameworks, Next.js uses `webpack` to accomplish this, which is the package responsible for generating the `Critical dependency` warning. Next.js reports this warning to developers in its console output. (Side note: my guess is that Prettier Playground does not have this bundler warning because `prettier` itself is not being bundled by `webpack` in the same way or at all.)
2. The effective meaning of `the request of a dependency is an expression` is that modules are being imported ["in such a way that webpack cannot reliably determine what code is used and which bits can be ignored."](https://medium.com/tomincode/hiding-critical-dependency-warnings-from-webpack-c76ccdb1f6c1) (Additional [source](https://github.com/i18next/next-i18next/issues/1545#issuecomment-976719749).) **Indeed, examining `prettier/index.mjs` reveals that there are multiple instances in which `require()` arguments are not `string`s or statically determinable `string`s.**
3. Adding `serverComponentsExternalPackages: ['prettier']` to `next.config.js` doesn't seem to have any effect when reproducing the issue in a route handler. The reason would be since, according to the [docs page](https://nextjs.org/docs/app/api-reference/next-config-js/serverComponentsExternalPackages) you linked to, `prettier` is "working on compatibility" and as such is automatically opted out of bundling triggered by an import from a server component or route handler.

It looks like others ([example](https://github.com/browserslist/browserslist/issues/199#issuecomment-350029083), [example](https://github.com/i18next/next-i18next/issues/1545#issuecomment-1005990731)) have found the following workaround does prevent the warning from being logged:
```mjs
// next.config.mjs
/** @type {import('next').NextConfig} */
const config = {
  // ...
  webpack: (config) => {
    config.module = {
      ...config.module,
      exprContextCritical: false // suppress warning caused within package 'prettier'
    };
    return config;
  }
  // ...
};
export default config;
```

**TL;DR**: The warning is caused by the `webpack`ing of `prettier` because `webpack` doesn't like that `prettier` uses dynamic arguments in `require()` statements.

So, to avoid this problem, one could either change the default Next.js behavior to prevent bundling like this or change `prettier` to use `require()`s that can be statically evaluated. I'm no subject matter expert, but the latter option makes more sense to me, as other meta-frameworks and build pipelines might want to run the `prettier` `npm` package through `webpack` as well in the future. That said, I am more than happy to open a discussion in the Next.js community about this if, with this additional info in mind, you still believe the issue is best resolved there. Thanks again!

**Comment by kachkaev** [comment]
Thanks for a very detailed investigation! 💯

Yeah I guess we’ll have to wait till at least one major release of Prettier to make any changes here. Prettier v3 is a step towards ESM, but the imports are a bit ‘unconventional’ to maintain compatibility with CommonJS. Hopefully, we won’t have to support CommonJS forever and a standard ESM import tree will become a not so distant reality!

**Comment by fisker** [comment]
Actually, we have test for such use case, https://github.com/prettier/prettier/blob/main/scripts/tools/bundle-test/index.js

I guess you expect prettier to work in browser? Then you should try to make it resolve to umd entry instead of the esm version.

```
 ⚠ ./node_modules/prettier/index.mjs
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Critical dependency: the request of a dependency is an expression
```

This is not expect at all.

**Comment by fisker** [comment]
I haven't checked how next.js configured, but something should be wrong for this option https://webpack.js.org/configuration/resolve/#resolveimportsfields

**Comment by LudoLogical** [comment]
Thank you both!
Fisker, I'll pass your insight along to the Next.js community today and see if we can't get this issue resolved on their end.
Alexander, yet again, your explanation was awesome; I understand completely. If I determine that the issue is indeed with prettier itself after consulting with the Next.js community, I'll reply here again and @ mention you so you have the full picture, but it sounds like the best path forward in the near to mid-term future is either resolving the issue in the Next.js repo or implementing a temporary workaround/warning suppression/etc. there.
